### PR TITLE
fix(linux_dns_issue): Re-trying DNS lookup of instance name after appending the domain from AG listener

### PR DESCRIPTION
**Pull request by AdityaNPL**
`Dns.GetHostEntry(instance)` fails intermittently on linux platforms if `instance` doesn't have the full domain

### CODE REVIEW

File: src/SmoFacade/Listener.cs [hunk 1]
```diff
@@ -134,15 +135,27 @@ private static string AgListenerName(string dataSource)
...
+            else if (parts.Length == 2)
+                // NamedInstances: chop instance name, resolve DNS, slap instance name back on!
+                connBuilder.DataSource = $"{ResolveDnsAndGetHostName(connBuilder.DataSource, parts[0])}\\{parts[1]}";
+            else
+                throw new ArgumentException($"agInstanceName param {agInstanceName} cannot be resolved by DNS");
 
       return new Server(connBuilder.ToString(), credentialName);
     }
+
+    private static string ResolveDnsAndGetHostName(string hostUrl, string instance)
```
Comment (Kaneraz) [inline src/SmoFacade/Listener.cs:149]: This looks like what @am99 was suggesting, but it's unclear here the difference between the instance and hostUrl. Let's rename and/or document the difference here.

**Review by Kaneraz — COMMENTED** [review]
Some odd spacing showing up in the diff, but it looks like it's doing what we set out to do.

File: src/SmoFacade/Listener.cs (hunk 1 above)
Comment (AdityaNPL) [reply src/SmoFacade/Listener.cs]: Sure have renamed and added some summary comments - let me know if you guys want the wording changed

File: src/SmoFacade/Listener.cs [hunk 2]
```diff
@@ -135,14 +135,33 @@ private static string AgListenerName(string dataSource)
...
+    /// <summary>
+    ///   Resolves "instanceName" to a FQDNS name
+    ///   However on Unix OS, "instanceName" should be a complete domain - "{sub domain}.{second level domain}.{...}.{top level domain}" (eg: "xyz.abc.def.com")
+    ///   Therefore, append from the {second-level} -> {top-level} domains from the "hostDomain" to the "instanceName" to make it a complete domain if it is not already
+    /// </summary>
+    /// <param name="hostDomain">The complete domain for the SQL server</param>
+    /// <param name="instanceName">The specific name of an instance within the server</param>
+    private static string ResolveDnsAndGetHostName(string hostDomain, string instanceName)
+    {
+      if (Environment.OSVersion.Platform == PlatformID.Unix && !instanceName.Contains('.'))
```
Comment (AdityaNPL) [inline src/SmoFacade/Listener.cs:157]: had to change 
from `RuntimeInformation.IsOSPlatform(OSPlatform.Linux)` 
to `Environment.OSVersion.Platform == PlatformID.Unix`
due to .Net Framework 4.7 compatibility issues

**Review by alexirion10 — COMMENTED** [review]
Code looks ok... but this is tricky. 
I'm not certain that `instanceName.Contains('.')` means that it's a FQDN necessarily
Did you test?  How did you test?  Can you provide some sample I/O?

**Comment by AdityaNPL** [comment]
I'm not sure if there is another way to check? I thought that if it has at least one `.` it could already be a FQDN?
Do we want to ignore if it is or not and just blindly append the later part of the domain from the `hostDomain` to the instance name extracted from `instanceName`?

i.e. if `hostDomain = "abc.def.ghi"` and `instanceName="uvw.xyz"` we do dns lookup on `uvw.def.ghi` ??

**Comment by alexirion10** [comment]
Exactly...
I don't have an answer to that. I'm asking if you've tried and thought through all of these possibilities

**Comment by AdityaNPL** [comment]
@alexirion10 checked where this function is being called from and it looks like the only values being passed in are the replica instance names - which won't be FQDN
Have removed the check as I feel we don't need it anymore

Also renamed a few names as I am understanding the terminologies and tech stack a little better now 😄 

What do you think?

File: src/SmoFacade/Listener.cs [hunk 3]
```diff
@@ -135,14 +135,52 @@ private static string AgListenerName(string dataSource)
...
+    /// </summary>
+    /// <param name="agListenerDomain">The complete domain for the SQL server AG listener</param>
+    /// <param name="agReplicaInstanceName">The specific name of a replica instance within the AG</param>
+    private static string ResolveDnsAndGetHostName(string agListenerDomain, string agReplicaInstanceName)
+    {
+      Tuple<string, string> domain_port = SplitDomainPort(agListenerDomain);
+      if (Environment.OSVersion.Platform == PlatformID.Unix)
+      {
+        string[] domainFragments = domain_port.Item1.Split('.');
+        domainFragments[0] = "";
```
Comment (alexirion10) [inline src/SmoFacade/Listener.cs:161]: Why are you setting domainFragments[0] to ""?  This seems random... would like a comment here or soemthing

File: src/SmoFacade/Listener.cs [hunk 4]
```diff
@@ -135,14 +135,52 @@ private static string AgListenerName(string dataSource)
...
+    /// <summary>
+    ///   Resolves "agReplicaInstanceName" to a FQDNS name
+    ///   However on Unix OS, "agReplicaInstanceName" should be a complete domain - "{sub domain}.{second level domain}.{...}.{top level domain}" (eg: "xyz.abc.def.com")
+    ///   Therefore, we use {second-level} -> {top-level} part of the domain from "agListenerDomain" and append it to "agReplicaInstanceName"
+    /// </summary>
+    /// <param name="agListenerDomain">The complete domain for the SQL server AG listener</param>
+    /// <param name="agReplicaInstanceName">The specific name of a replica instance within the AG</param>
+    private static string ResolveDnsAndGetHostName(string agListenerDomain, string agReplicaInstanceName)
+    {
+      Tuple<string, string> domain_port = SplitDomainPort(agListenerDomain);
```
Comment (alexirion10) [inline src/SmoFacade/Listener.cs:157]: no snake_case please! we use camelCase everywhere else in the code

**Review by alexirion10 — CHANGES_REQUESTED** [review]

**Comment by Kaneraz** [comment]
@AdityaNPL if we're concerned about getting the FQDN instead of just a hostname we can attempt to resolve the hostname first and if it fails, then attempt to append from the original listener name.

**Comment by AdityaNPL** [comment]
Changing this to Draft as I think this needs much more refactoring before it is ready to release publicly.
Leaving PR open to continue with reviews and suggestions.

Sure @Kaneraz , yes I think changing to a try-catch might be less error prone for us right now - will change to that

File: src/SmoFacade/Listener.cs (hunk 3 above)
Comment (AdityaNPL) [reply src/SmoFacade/Listener.cs:161]: sure will put in a comment there

It's to strip off the first part of the domain before appending it to the instance name
so if it's like `abc.def.ghi` we want only `.def.ghi` to be appended to the instance name

I think we should end up using `UriBuilder` class in the end rather than our own string manipulations:
https://docs.microsoft.com/en-us/dotnet/api/system.uribuilder?view=netcore-2.0

File: src/SmoFacade/Listener.cs [hunk 5]
```diff
@@ -135,14 +135,52 @@ private static string AgListenerName(string dataSource)
...
+      if (Environment.OSVersion.Platform == PlatformID.Unix)
+      {
+        string[] domainFragments = domain_port.Item1.Split('.');
+        domainFragments[0] = "";
+        agReplicaInstanceName += string.Join(".", domainFragments);
+      }
+      return Dns.GetHostEntry(agReplicaInstanceName).HostName+domain_port.Item2;
+
+    }
+    private static Tuple<string,string> SplitDomainPort(string domainAndPort)
```
[the remark this answers is not in the document]
Comment (AdityaNPL) [reply src/SmoFacade/Listener.cs:167]: looks like in c# I can actually create a tuple (like python) and I can access it via the names directly

It's called tuple types - what do you think about these?
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/value-tuples

File: src/SmoFacade/Listener.cs [hunk 6]
```diff
@@ -133,16 +133,15 @@ private static string AgListenerName(string dataSource)
     private static Server AgListenerNameToServer(ref SqlConnectionStringBuilder connBuilder, string agInstanceName,
       string credentialName)
     {
-      var parts = agInstanceName.Split('\\');
-      if(parts.Length == 1)
```
Comment (AdityaNPL) [inline src/SmoFacade/Listener.cs:137]: `SplitDomainPort()` will now handle this

File: src/SmoFacade/Listener.cs [hunk 7]
```diff
@@ -154,33 +153,47 @@ private static string AgListenerName(string dataSource)
...
+      {
+        Console.WriteLine($"Unable to get dns for instance '{instanceDomain}'\n{e}");
+
+        // Re-try with full domain appended to the instanceDomain by stripping off domain from listner and appending to instance
+        // eg: if listner is "abc.def.ghi" we want to append ".def.ghi" to the instance
+        string[] domainFragments = listenerDomain.Split('.');
+        domainFragments[0] = "";
+        string instanceDomainFull = $"{instanceDomain}{string.Join(".", domainFragments)}";
 
+        Console.WriteLine($"Retrying with full domain '{instanceDomainFull}");
```
Comment (AdityaNPL) [inline src/SmoFacade/Listener.cs:175]: Do we want these logging messages in the final release - I think they would be helpful for debugging for anyone using this

What do you guys think?

File: src/SmoFacade/Listener.cs [hunk 8]
```diff
@@ -135,14 +135,52 @@ private static string AgListenerName(string dataSource)
...
+      {
+        string[] fragments = domainAndPort.Split(new char[] {','}, 2);
+        domian = fragments[0];
+        port = ","+fragments[1];
+      }
+      else if (domainAndPort.Contains('\\'))
+      {
+        string[] fragments = domainAndPort.Split(new char[] {'\\'}, 2);
+        domian = fragments[0];
+        port = "\\"+fragments[1];
```
[the remark this answers is not in the document]
Comment (AdityaNPL) [reply src/SmoFacade/Listener.cs]: done

File: src/SmoFacade/Listener.cs [hunk 9]
```diff
@@ -133,16 +133,67 @@ private static string AgListenerName(string dataSource)
...
+        domain = fragments[0];
+        port = $",{fragments[1]}";
+      }
+      else if (domainAndPort.Contains('\\')) // handling named instances the same way as ports
+      {
+        string[] fragments = domainAndPort.Split(new char[] {'\\'}, 2);
+        domain = fragments[0];
+        port = $"\\{fragments[1]}";
+      }
+      return (domain, port);
```
Comment (alexirion10) [inline src/SmoFacade/Listener.cs:196]: This method needs unit tests

```C#
var splitChar = domainAndPort.Contains(',') ? ',' : domainAndPort.Contains('\\') ? ' \\' : null;
if(splitChar != null)
{
  string[] fragments = domainAndPort.Split(splitChar, 2);
  domain = fragments[0];
  port = $"{splitChar}{fragments[1]}";
}

return (domain, port);

```

File: src/SmoFacade/Listener.cs (hunk 7 above)
Comment (alexirion10) [reply src/SmoFacade/Listener.cs:175]: Console logging isn't desireable in a Nuget package release IMO.
If you want log statements, they should go through Microsoft ILogger, for which there are many extensions to... like nlog
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-5.0

File: src/SmoFacade/Listener.cs (hunk 9 above)
Comment (AdityaNPL) [reply src/SmoFacade/Listener.cs:196]: ^^ ah yes, much better way of doing it - thanks! 😊 

Yes agreed, there can be many edge cases here - will look at creating some unit tests

Was just implementing the PR suggestions from above. 
For next step, as mentioned before, I'm going to look at using this class instead of doing the fragmentations myself: 
[`UriBuilder`](https://docs.microsoft.com/en-us/dotnet/api/system.uribuilder?view=netcore-2.0)

Comment (AdityaNPL) [reply src/SmoFacade/Listener.cs:196]: For `UriBuilder`, looks like it expects a well formatted URI (with protocol), which we don't have.
Plus I can't see functions there to segment out the domain name Itself.
(for ports, we are also handling named instances - which is SQL Sever specific so it won't be able to handle those either)

reverting back to using normal string split functions.

File: src/SmoFacade/Listener.cs (hunk 7 above)
Comment (AdityaNPL) [reply src/SmoFacade/Listener.cs:175]: Had a look at `ILogger` and looks like it might be better to implement that (if we agree to do so) in it's own PR. There are many components that we will need to create for it to function (Logger models, categories, logging levels, etc.)

Shelfing for now, will come back to it if we feel it is necessary

File: src/SmoFacade/Listener.cs [hunk 10]
```diff
@@ -130,19 +130,70 @@ private static string AgListenerName(string dataSource)
...
+    /// <param name="agReplicaInstanceName"> The name for an instance within the AG (for which we are trying to get the FQDN)</param>
+    /// <param name="agListenerDomain"> The complete domain for the AG listener</param>
+    private static string ResolveDnsHostNameForInstance(string agReplicaInstanceName, string agListenerDomain)
+    {
+      // Sometimes instances and listeners have ports(or "named instances")
+      // Therefore, we strip them off before calling DNS.GetHostEntry() and then add them back to the result 
+      var (listenerDomain, listenerPort) = SplitDomainAndPort(agListenerDomain);
+      var (instanceName, instancePort) = SplitDomainAndPort(agReplicaInstanceName);
+      // preference is to add back 'instancePort' over 'listenerPort' (in almost all cases they should be identical)
+      var port = instancePort ?? listenerPort;
```
Comment (AdityaNPL) [inline src/SmoFacade/Listener.cs:161]: Had a chat with Chris and we should look to change priority here - we want to prefer the port over named instance.
Mainly because ports are TCP/IP standards while named instances are only SQL Server standards. 

What do you guys think?

(should also rename `instancePort` to `instancePortOrNamedInstance` so that it is clear 😅 )

File: src/SmoFacade/Listener.cs [hunk 11]
```diff
@@ -130,19 +130,70 @@ private static string AgListenerName(string dataSource)
...
+        listenerDomainFragments[0] = null;
+        var instanceDomain = $"{instanceName}{string.Join(".", listenerDomainFragments)}";
+
+        return $"{Dns.GetHostEntry(instanceDomain).HostName}{port}";
+      }
     }
+
+    // This function handles named instances ("<domain>\<named instance>") in the same way as ports
+    private static (string, string) SplitDomainAndPort(string domainAndPort)
+    {
```
Comment (AdityaNPL) [inline src/SmoFacade/Listener.cs:182]: @alexirion10 working on Unit tests for this - was wondering if we should have some for `ResolveDnsHostNameForInstance()` too to check it retries correctly?

Or is just checking this one fine?

Comment (alexirion10) [reply src/SmoFacade/Listener.cs:182]: the more tests the better

**Comment by AdityaNPL** [comment]
Have refactored and added tests for the utility functions - I think should be ready for release
@alexirion10 / @Kaneraz can you pls give it a look over and let me know if you have any suggestions

File: src/SmoFacade/Listener.cs [hunk 12]
```diff
@@ -130,19 +131,81 @@ private static string AgListenerName(string dataSource)
...
+    ///   Therefore, if dns lookup on just the instance name fails, we retry after appending the domain fragments from the listener to the instance name
+    /// </summary>
+    /// <param name="agReplicaInstanceName"> The name for an instance within the AG (for which we are trying to get the FQDN)</param>
+    /// <param name="agListenerDomain"> The complete domain for the AG listener</param>
+    private static string ResolveDnsHostNameForInstance(string agReplicaInstanceName, string agListenerDomain)
+    {
+      // Sometimes instances and listeners have ports or "named instances" (NI)
+      // Therefore, we strip them off before calling DNS.GetHostEntry() and then add them back to the result 
+      var (listenerDomain, listenerPortOrNi) = SplitDomainAndPort(agListenerDomain);
+      var (instanceName, instancePortOrNi) = SplitDomainAndPort(agReplicaInstanceName);
```
Comment (alexirion10) [inline src/SmoFacade/Listener.cs:160]: I'm not big on acronyms in variable names... `Ni` --> `NamedInstance` is verbose but more readable

File: tests/AgDatabaseMove.Unit/SmoFacadeListenerTest.cs [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+    // {input, expectedDomain, expectedPort (can be null)}
+    public static IEnumerable<object[]> ValidPorts => new List<object[]> {
+      new object[] { $"{DOMAIN}", DOMAIN },
+      new object[] { $"{DOMAIN},123", DOMAIN, ",123" },
+      new object[] { $"{DOMAIN}\\SQL", DOMAIN, "\\SQL" },
+
+      new object[] { $"{HOST}", HOST },
+      new object[] { $"{HOST},123", HOST, ",123" },
+      new object[] { $"{HOST}\\SQL", HOST, "\\SQL" }
```
Comment (alexirion10) [inline tests/AgDatabaseMove.Unit/SmoFacadeListenerTest.cs:24]: No need to interpolate $"{DOMAIN}" and $"{HOST} when it's already a string and not being concatenated.
Some other strings like ",123" and "\\SQL" could be put in const string variables too

File: tests/AgDatabaseMove.Unit/SmoFacadeListenerTest.cs [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+    };
+
+    [Theory]
+    [MemberData(nameof(ValidPorts))]
```
Comment (alexirion10) [inline tests/AgDatabaseMove.Unit/SmoFacadeListenerTest.cs:28]: `[InlineData]` is a bit more simple if you're only going to use the Enumerable inputs once
https://hamidmosalla.com/2017/02/25/xunit-theory-working-with-inlinedata-memberdata-classdata/

**Review by alexirion10 — COMMENTED** [review]
I mean... I'm sure it works, but man this code is messy and fragile. I'm nervous with the amount of string parsing happening in here

File: src/SmoFacade/Listener.cs (hunk 12 above)
Comment (AdityaNPL) [reply src/SmoFacade/Listener.cs:160]: I was debating this myself, it looked very long - so thought it might be better to name it `Ni` and put in a comment above
`// Sometimes instances and listeners have ports or "named instances" (NI)`

But yes, I can see your point and I now agree too - should go for more easy to read names

File: tests/AgDatabaseMove.Unit/SmoFacadeListenerTest.cs (hunk 14 above)
Comment (AdityaNPL) [reply tests/AgDatabaseMove.Unit/SmoFacadeListenerTest.cs:28]: Looks like we can't use `[InlineData]` for the first 2 tests - it doesn't support interpolated strings
Will change the last one

**Comment by AdityaNPL** [comment]
The code logic of our fix:
1. Strip off any ports/named instances from both the listener and the instance
2. Try to resolve to FQDN by just using the instance name
3. If it fails (as it does sometimes on linux) we append the domain fragments from the listener to the instance name (to make it a complete domain) and then re-try
4. To the resolved FQDN, add back a port or a named instance (if available from step 1)

@alexirion10 I agree, but looking above at the steps we need to do, I think the string manipulations are required. 
I did have a look at using some in-built c# classes to perform these actions as well (like `UriBuilder` and `SqlConnectionStringBuilder`) but couldn't find any that could

---

I have some suggestions that could give us more confidence:
- I can intensify testing a bit more?
  - For example: refactor the string manipulation in `ResolveDnsHostNameForInstance()` into its own function and write some tests for that?
- We can also add some string format checks to make sure inputs are in correct format before we start splitting and slicing?
- We could go down the path of `regex` - but I opted not to because it is usually more complex to develop and felt it might introduce more bugs for something we can do with some simple string manipulations (which are quicker to implement and easier to follow) 

What are your thoughts?

**Review by Kaneraz — APPROVED** [review]
As we've discussed this is a bit of a tangle, but after looking at it a few time I'm not seeing a clear way to untangle it. Once we get things in place to resolve WINS on the containers this runs in, we can consider removing the hack.

Thanks for the unit tests, they give me confidence that this is going to remain stable as we keep moving forward and balances out my concerns with how tangled it feels.

File: src/SmoFacade/Listener.cs [hunk 15]
```diff
@@ -133,16 +133,67 @@ private static string AgListenerName(string dataSource)
...
+        return new Server(connBuilder.ToString(), credentialName);
+      }
+      catch (Exception e)
+      {
+        throw new ArgumentException($"agInstanceName param {agInstanceName} cannot be resolved by DNS", e);
+      }
+    }
+
+    /// <summary>
+    ///   Resolves "agReplicaInstanceName" to a FQDNS name
```
[the remark this answers is not in the document]
Comment (AdityaNPL) [reply src/SmoFacade/Listener.cs]: looks like this is already fixed
(this is from an old commit)

File: src/SmoFacade/Listener.cs [hunk 16]
```diff
@@ -1,3 +1,4 @@
+[assembly: System.Runtime.CompilerServices.InternalsVisibleTo("AgDatabaseMove.Unit")]
```
[the remark this answers is not in the document]
Comment (AdityaNPL) [reply src/SmoFacade/Listener.cs:1]: PR: https://github.com/factset/AgDatabaseMove/pull/63

File: src/SmoFacade/Listener.cs [hunk 17]
```diff
@@ -133,16 +133,67 @@ private static string AgListenerName(string dataSource)
...
+      }
+      catch (Exception e)
+      {
+        throw new ArgumentException($"agInstanceName param {agInstanceName} cannot be resolved by DNS", e);
+      }
+    }
+
+    /// <summary>
+    ///   Resolves "agReplicaInstanceName" to a FQDNS name
+    ///   However on Unix OS, "agReplicaInstanceName" should be a complete domain - "{sub domain}.{second level domain}.{...}.{top level domain}" (eg: "xyz.abc.def.com")
```
[the remark this answers is not in the document]
Comment (AdityaNPL) [reply src/SmoFacade/Listener.cs]: PR: https://github.com/factset/AgDatabaseMove/pull/63
